design-document.md

```diff
@@ -14,4 +24,210 @@
 
 ## Open questions and system parameters
 
-## Cryptographic dependencies
+## Cryptographic components
+
+### The input-validation protocol
+
+**Notation.**
+Fix a finite field K. We let K^n denote the set of vectors over K of length n.
+We write the linear, s-way secret sharing of an element x of K^n as {x:1}, ...,
+{x:s}, where {x:i} is the share held by the i-th party. We write {x} as
+shorthand for the sequence {x:1}, ..., {x:s}.
+
+Prio combines standard linear secret sharing with a new type of
+probabilistically checkable proof (PCP) system, called a fully linear PCP. The
+input-input validation protocol can be described in terms of three main
+algorithms:
+
+1. pf := Prove(x) denotes generation of a proof pf of the validity of input x.
+   This algorithm is executed by the user.
+1. {vf:i} := Query({x:i}, {pf:i}, r) denotes computation of the verification
+   share {vf:i} for input share {x:i} and proof share {pf:i}. This algorithm is
+   executed by each of the aggregators; input r denotes the joint randomness
+   shared by all of the aggregators.
+1. b := Decide({vf}, r) denotes the execution of the decision procedure on input
+   shares {vf} and joint randomness r. The output b is a boolean indicating
+   whether the input is deemed valid. This algorithm is run by the leader.
+
+The values above have following types:
+
+1. Input x is an element of K^n for some integer n.
+1. Proof pf is an element of K^p(n) for some function p.
+1. The joint randomness r is an element of K^u(n) for some function u.
+1. Each verification share {vf:i} is an element of K^v(n) for some function v.
+
+Before the protocol begins, the aggregators agree on joint randomness r and
+designate one of the aggregators as the leader. The protocol proceeds as
+follows:
+
+1. The user runs pf := Prove(x). It splits x and pf into {x} and {pf}
+   respectively and sends ({x:i}, {pf:i}) to aggregator i.
+1. Each aggregator i runs {vf:i} := Query({x:i}, {pf:i}, r) ands sends {vf:i} to
+   the leader.
+1. The leader runs b := Decide({vf}, r) and sends b to each of the aggregators.
+
+If b=1, then each aggregator i adds its input share {x:i} into its share of the
+aggregate. Once all of the inputs have been validated and aggregated, the
+aggregators send their aggregate shares to the leader, who adds them together to
```

Comment:
I think this should say something like "Once all of the inputs have been validated and a sufficient number of input shares have been aggregated, the aggregators send their aggregate shares to the leader...", to make it clear that aggregators aren't disclosing individual input shares to the leader.

Reply:
Changed to "Once a sufficient number of shares have been ..."